File: Assets/Scripts/App/Lua/App/Manager/Spawner.lua
require('App/Generic/SpawnMonsterData')
require('App/Object/Monster')
require('App/Generic/PrizeCell')
require('App/Generic/ObstaclePrize')
require('App/Object/BossMonster')

Spawner = Object:SubClass('Spawner')

Spawner.cs = nil
Spawner.mono = nil
Spawner.sceneManager = nil
Spawner.gameManager = nil

Spawner.fireTarget = nil
Spawner.signTrf = nil

Spawner.spawnedComplete = false
Spawner.isWaveInProgress = false
Spawner.isStarted = false
Spawner.isPaused = true
Spawner.currentWaveIndex = 0
Spawner.currentWaveMonsterIndex = 0
Spawner.waveTimer = 0
Spawner.monsterSpawnTimer = 0

Spawner.pathList = nil
Spawner.waveDataList = nil
Spawner.obstacleList = nil
Spawner.prizeTowerList = nil

Spawner.obstaclePrizeList = nil

Spawner.spawnedMonsterList = nil
Spawner.spawnedTowerList = nil
Spawner.spawnedObstacleList = nil
Spawner.spawnedBoss = nil

Spawner.nowWaveSpawnList = nil

function Spawner:Construct()
    self.gameManager = GameManager.cs
    self.sceneManager = GameManager.cs.sceneManager

    local obj = Instantiate(Resources.Load('Prefabs/Spawner'))
    Destroy(obj:GetComponent('Spawner'))
    self.cs = LuaSpawner()
    self.mono = obj:AddComponent(typeof(MonoScript))

    self:InitAction()

    self.nowWaveSpawnList = List()
    self.spawnedMonsterList = CS.System.Collections.Generic.List(IMonster)()
    self.spawnedTowerList = CS.System.Collections.Generic.List(ITower)()
    self.spawnedObstacleList = CS.System.Collections.Generic.List(IObstacle)()
    self.obstaclePrizeList = List()

    self.signTrf = self.mono.transform:Find('CollectingFiresSign')
end

function Spawner:InitAction()
    -- monoScript
    self.mono.onUpdateAction = function()
        self:Update()
    end
    -- LuaSpawner
    self.cs.onGetCarrotAction = function()
        return self:GetCarrot()
    end
    self.cs.onPushAllGameObjectAction = function()
        self:OnPushAllGameObject()
    end
    self.cs.onWinJudgeAction = function()
        return self:WinJudge()
    end
    self.cs.onPauseWavesAction = function()
        self:PauseWaves()
    end
    self.cs.onResumeWavesAction = function()
        self:ResumeWaves()
    end
    self.cs.onStartSpawnAction = function()
        self:StartSpawn()
    end
    self.cs.onGetAllMonstersAction = function()
        return self:GetAllMonsters()
    end
    self.cs.onGetCollectingFiresTargetAction = function()
        return self:GetCollectingFiresTarget()
    end
    self.cs.onSetCollectingFiresAction = function(m)
        self:SetCollectingFires(m)
    end
    self.cs.onCancelCollectingFiresTargetAction = function()
        self:CancelCollectingFiresTarget()
    end
    self.cs.onInitAction = function(mapData)
        self:Init(mapData)
    end
    self.cs.onCreateTowerObjectAction = function(towerData, v3)
        self:CreateTowerObject(towerData, v3)
    end
    self.cs.onUpGradeTowerAction = function(v3)
        self:UpGradeTower(v3)
    end
    self.cs.onSellTowerAction = function(v3)
        self:SellTower(v3)
    end
    self.cs.onGetNowWaveCountAction = function()
        return self:GetNowWaveCount()
    end
end

-- 外部获取变量
function Spawner:GetCarrot()
    return nil
end

function Spawner:GetNowWaveCount()
    return 0
end

function Spawner:GetAllMonsters()
    return self.spawnedMonsterList
end

-- unity回调
function Spawner:Update()
    if self.isStarted == false or self.isPaused == true then
        return
    end

    if self.currentWaveIndex < self.waveDataList.Count then
        if self.isWaveInProgress == false and self.waveTimer >= self.waveDataList[self.currentWaveIndex].waveDuration then
            self:StartNewWave()
        end

        if self.currentWaveMonsterIndex >= self.nowWaveSpawnList.count then
            self.waveTimer = self.waveTimer + Time.deltaTime
            if self.isWaveInProgress == true then
                self:EndCurrentWave()
            end
        else
            self:HandleMonsterSpawning()
        end
    else
        self.spawnedComplete = true
        if self.spawnedBoss == nil then
            self:CreateBossMonster()
        end
    end

    for i = 0, self.obstaclePrizeList.count - 1 do
        local obstaclePrize = self.obstaclePrizeList:Get(i)
        if obstaclePrize.obstacle.IsDead == true and obstaclePrize:HasPrize() == true then
            -- 生成奖励
            self:CreatePrize(obstaclePrize)
            self.obstaclePrizeList:DelayRemove(obstaclePrize)
        end
    end
    self.obstaclePrizeList:StartRemove()
end

function Spawner:CreateBossMonster()
    local bossObj =
        Instantiate(Resources.Load('AB/Object/Boss/BossMonster' .. GameManager.sceneManager:GetLevelID() - 1))
    local boss = BossMonster(bossObj)

    boss.obj.transform:SetParent(self.cs.transform)
    boss.obj.transform.localScale = Vector3.one
    -- 添加进列表
    self.spawnedMonsterList:Add(boss.cs)
    self.spawnedBoss = boss
    -- 加载怪物数据
    local monsterData = Resources.Load('AB/Data/BossMonsterData/Boss' .. GameManager.sceneManager:GetLevelID() - 1)
    boss:Init(self.pathList, 1, monsterData)
end

function Spawner:CreatePrize(obstaclePrize)
    for i = 0, obstaclePrize.prizeCellList.count - 1 do
        local prizeCell = obstaclePrize.prizeCellList:Get(i)

        -- 加载towerData
        local towerDataMap = Resources.Load('Data/Tower/TowerMap')
        local towerData = towerDataMap:GetData(prizeCell.prizeTowerType)
        local pos = CSMap.GetCellCenterPos(prizeCell.cell)
        local towerObj = self.gameManager.poolManager:GetObject(towerData.prefabsPath)
        local tower = towerObj:GetComponent(typeof(ITower))
        towerObj.transform:SetParent(self.mono.transform)
        towerObj.transform.localScale = Vector3.one
        towerObj.transform.position = pos

        CSMap.GetCell(pos).tower = tower
        self.spawnedTowerList:Add(tower)
    end
end

function Spawner:Init(mapData)
    self.spawnedComplete = false
    self.isWaveInProgress = false
    self.isStarted = false
    self.isPaused = true
    self.currentWaveIndex = 0
    self.currentWaveMonsterIndex = 0
    self.waveTimer = 0
    self.monsterSpawnTimer = 0

    self.pathList = mapData:GetPathList()
    self.waveDataList = mapData:GetWaveData()
    self.obstacleList = mapData:GetObstacle()
    self.prizeTowerList = mapData.towerList

    self:CreateObstacles()
end

-- 出怪逻辑
function Spawner:StartNewWave()
    self.isWaveInProgress = true
    self.currentWaveMonsterIndex = 0
    self.waveTimer = 0
    self.monsterSpawnTimer = 0

    self.nowWaveSpawnList:Clear()

    local list = self.waveDataList[self.currentWaveIndex].eachWaveDataList
    for i = 0, list.Count - 1 do
        local e = list[i]
        for j = 0, e.monsterCount - 1 do
            local d = SpawnMonsterData:New(e.monsterType, e.monsterDuration, e.hard)
            self.nowWaveSpawnList:Add(d)
        end
    end
end

function Spawner:EndCurrentWave()
    self.isWaveInProgress = false
    self.currentWaveIndex = self.currentWaveIndex + 1
end

function Spawner:HandleMonsterSpawning()
    if self.currentWaveIndex == 0 and self.currentWaveMonsterIndex == 0 then
        local type = self.nowWaveSpawnList:Get(self.currentWaveMonsterIndex).monsterType
        local hard = self.nowWaveSpawnList:Get(self.currentWaveMonsterIndex).hard
        self:SpawnerMonster(type, hard)
        self.currentWaveMonsterIndex = self.currentWaveMonsterIndex + 1
        self.monsterSpawnTimer = 0
    else
        self.monsterSpawnTimer = self.monsterSpawnTimer + Time.deltaTime
        if self.monsterSpawnTimer >= self.nowWaveSpawnList:Get(self.currentWaveMonsterIndex).nextSpawnTime then
            local type = self.nowWaveSpawnList:Get(self.currentWaveMonsterIndex).monsterType
            local hard = self.nowWaveSpawnList:Get(self.currentWaveMonsterIndex).hard
            self:SpawnerMonster(type, hard)
            self.currentWaveMonsterIndex = self.currentWaveMonsterIndex + 1
            self.monsterSpawnTimer = 0
        end
    end
end

function Spawner:SpawnerMonster(type, hard)
    -- 获取预设体
    local monsterObj = self.gameManager.poolManager:GetObject('Object/Monster/' .. type:ToString())
    DestroyImmediate(monsterObj:GetComponent('Monster'), true)
    -- 实例化并进行lua初始化
    local monster = Monster(monsterObj)
    -- 设置位置
    monster.obj.transform:SetParent(self.cs.transform)
    monster.obj.transform.localScale = Vector3.one
    -- 添加进列表
    self.spawnedMonsterList:Add(monster.cs)
    -- 加载怪物数据
    local monsterDataMap = Resources.Load('Data/Monster/MonsterDataMap')
    if monsterDataMap == nil then
        monsterDataMap = Resources.Load('AB/Data/')
    end
    local monsterData = monsterDataMap:GetData(type)
    monster:Init(self.pathList, hard, monsterData)
end

-- 开始、暂停、继续
function Spawner:StartSpawn()
    self.isStarted = true
    self.isPaused = false
    self.currentWaveIndex = 0
    self.currentWaveMonsterIndex = 0
    self.waveTimer = 0
    self.monsterSpawnTimer = 0
    self.isWaveInProgress = false
end

function Spawner:PauseWaves()
    self.isPaused = true
end

function Spawner:ResumeWaves()
    self.isPaused = false
end

-- 塔升级出售
function Spawner:UpGradeTower(pos)
    local tower = CSMap.GetCell(pos).tower
    if tower == nil then
        return
    end
    local level = tower:GetLevel()
    local data = tower:GetData()

    if level == 2 then
        return
    end

    if self.sceneManager:GetMoney() < data.prices[level + 1] then
        return
    end

    tower:UpGrade()

    self.sceneManager:UpdateMoney(-data.prices[level + 1])
    GameFacade.Instance:SendNotification('HIDE_BUILT_PANEL')
end

function Spawner:SellTower(pos)
    local cell = CSMap.GetCell(pos)
    local tower = cell.tower
    if tower == nil then
        return
    end
    local level = tower:GetLevel()
    local data = tower:GetData()

    self.sceneManager:UpdateMoney(data.sellPrices[level])
    -- 回收对象
    self.gameManager.poolManager:PushObject(tower.gameObject)
    self.spawnedTowerList:Remove(tower)
    cell.tower = nil
    GameFacade.Instance:SendNotification('HIDE_BUILT_PANEL')
end

function Spawner:CreateTowerObject(towerData, pos)
    if self.sceneManager:GetMoney() <= towerData.prices[0] then
        return
    end
    local towerObj = self.gameManager.poolManager:GetObject(towerData.prefabsPath)
    local tower = towerObj:GetComponent(typeof(ITower))
    towerObj.transform:SetParent(self.mono.transform)
    towerObj.transform.localScale = Vector3.one
    towerObj.transform.position = pos

    self.sceneManager:UpdateMoney(-towerData.prices[0])
    CSMap.GetCell(pos).tower = tower
    GameFacade.Instance:SendNotification('HIDE_BUILT_PANEL')
    self.spawnedTowerList:Add(tower)
end

function Spawner:CreateObstacles()
    for i = 0, self.obstacleList.Count - 1 do
        local name = self.obstacleList[i].obstacleName
        if name ~= 'None' then
            local cell = self.obstacleList[i]
            local obj = self.gameManager.poolManager:GetObject('Object/Obstacle/' .. cell.obstacleName)
            obj.transform:SetParent(self.cs.transform)
            obj.transform.localScale = Vector3.one
            obj.transform.position = CSMap.GetCellCenterPos(cell)
            cell.obstacle = obj
            self.spawnedObstacleList:Add(obj:GetComponent(typeof(IObstacle)))
            -- 绑定摧毁奖励
            -- 判断是否有奖励在对应方格中
            local obstaclePrize = ObstaclePrize(obj:GetComponent(typeof(IObstacle)))

            if string.find(name, '1X1') then
                -- 1X1
                for i = 0, self.prizeTowerList.Count - 1 do
                    local prizeTower = self.prizeTowerList[i]

                    if prizeTower.x == cell.X and prizeTower.y == cell.Y then
                        local prizeCell = PrizeCell(cell, prizeTower.towerType)
                        obstaclePrize:AddPrizeCell(prizeCell)
                    end
                end
            elseif string.find(name, '1X2') then
                -- 1X2
                for i = 0, self.prizeTowerList.Count - 1 do
                    local prizeTower = self.prizeTowerList[i]

                    if prizeTower.x == cell.X and prizeTower.y == cell.Y then
                        local prizeCell = PrizeCell(cell, prizeTower.towerType)
                        obstaclePrize:AddPrizeCell(prizeCell)
                    end
                    if prizeTower.x == cell.X + 1 and prizeTower.y == cell.Y then
                        local prizeCell = PrizeCell(Cell(Point(cell.X + 1, cell.Y)), prizeTower.towerType)
                        obstaclePrize:AddPrizeCell(prizeCell)
                    end
                end
            elseif string.find(name, '4X4') then
                -- 4X4
                for i = 0, self.prizeTowerList.Count - 1 do
                    local prizeTower = self.prizeTowerList[i]

                    if prizeTower.x == cell.X and prizeTower.y == cell.Y then
                        local prizeCell = PrizeCell(cell, prizeTower.towerType)
                        obstaclePrize:AddPrizeCell(prizeCell)
                    end
                    if prizeTower.x == cell.X + 1 and prizeTower.y == cell.Y then
                        local prizeCell = PrizeCell(Cell(Point(cell.X + 1, cell.Y)), prizeTower.towerType)
                        obstaclePrize:AddPrizeCell(prizeCell)
                    end
                    if prizeTower.x == cell.X and prizeTower.y == cell.Y + 1 then
                        local prizeCell = PrizeCell(Cell(Point(cell.X, cell.Y + 1)), prizeTower.towerType)
                        obstaclePrize:AddPrizeCell(prizeCell)
                    end
                    if prizeTower.x == cell.X + 1 and prizeTower.y == cell.Y + 1 then
                        local prizeCell = PrizeCell(Cell(Point(cell.X + 1, cell.Y + 1)), prizeTower.towerType)
                        obstaclePrize:AddPrizeCell(prizeCell)
                    end
                end
            end

            self.obstaclePrizeList:Add(obstaclePrize)
        end
    end
end

-- 集火
function Spawner:SetCollectingFires(monster)
    self.fireTarget = monster
    self.signTrf.gameObject:SetActive(true)
    self.signTrf:SetParent(monster:GetSignFather())
    self.signTrf.localPosition = Vector3.zero
    self.signTrf.localScale = Vector3.one
end
function Spawner:GetCollectingFiresTarget()
    return self.fireTarget
end

function Spawner:CancelCollectingFiresTarget()
    self.fireTarget = nil
    self.signTrf.gameObject:SetActive(false)
    self.signTrf:SetParent(self.mono.transform)
end

-- 缓存池
function Spawner:OnPushAllGameObject()
    self.obstaclePrizeList:Clear()

    self.signTrf.gameObject:SetActive(false)
    self:OnPushAllMonsters()
    self:OnPushAllObstacles()
    self:OnPushAllTowers()
end

function Spawner:OnPushAllMonsters()
    for i = 0, self.spawnedMonsterList.Count - 1 do
        self.gameManager.poolManager:PushObject(self.spawnedMonsterList[i].Transform.gameObject)
    end
    self.spawnedMonsterList:Clear()
end

function Spawner:OnPushAllTowers()
    for i = 0, self.spawnedTowerList.Count - 1 do
        self.gameManager.poolManager:PushObject(self.spawnedTowerList[i].Transform.gameObject)
    end
    self.spawnedTowerList:Clear()
end

function Spawner:OnPushAllObstacles()
    for i = 0, self.spawnedObstacleList.Count - 1 do
        self.gameManager.poolManager:PushObject(self.spawnedObstacleList[i].Transform.gameObject)
    end
    self.spawnedObstacleList:Clear()
end

-- Win
function Spawner:WinJudge()
    -- 出怪完成
    if self.spawnedComplete == false then
        return false
    end

    -- 所有怪物死亡
    for i = 0, self.spawnedMonsterList.Count - 1 do
        local m = self.spawnedMonsterList[i]
        if m.IsDead == false then
            return false
        end
    end

    return true
end

File: Assets/Scripts/App/Game/Spawner/Spawner.cs
using System.Collections.Generic;
using App.Data.DataClass;
using App.Data.DataClass.Game.Object;
using App.Data.DataClass.Map;
using App.Game.Generic.BaseObject;
using App.Game.Generic.Map;
using App.Game.Object.Carrot;
using App.Game.Object.Map;
using App.Game.Object.Monster;
using App.Game.Object.Obstacle;
using App.Game.Object.Tower;
using App.Game.SceneManager.NormalGame.interf;
using App.Static;
using App.Static.Enum;
using UnityEngine;

namespace App.Game.Spawner
{
    /// <summary>
    /// 对象生成器
    /// </summary>
    public class Spawner : MonoBehaviour, ISpawner
    {
        public Carrot carrot; // 萝卜
        public Transform startPoint; // 开始路牌位置
        private List<IMonster> monsters = new List<IMonster>(); // 已经出生的怪物
        private List<ITower> towers = new List<ITower>(); // 已创建的塔
        private List<IObstacle> obstaclesList = new List<IObstacle>(); // 已创建的障碍物

        private IMonster collectingFiresTarget; // 集火目标
        public Transform signTrans; // 集火标志

        private List<Cell> pathList;
        private List<Cell> obstacleList;

        #region 出怪

        public bool spawnedComplete; // 完成出怪
        private bool isWaveInProgress;
        private bool isStarted;
        private bool isPaused;
        private int currentWaveIndex;
        private int currentWaveMonsterIndex;
        private float waveTimer;
        private float monsterSpawnTimer;
        public List<WaveData> waveDataList = new List<WaveData>();
        private List<SpawnMonsterData> nowWaveSpawnList = new List<SpawnMonsterData>();

        #endregion

        private INormalSceneManager sceneManger;

        public virtual Carrot Carrot
        {
            get => carrot;
        }

        protected virtual void Update()
        {
            if (!isStarted || isPaused)
            {
                return;
            }

            if (currentWaveIndex < waveDataList.Count)
            {
                if (!isWaveInProgress && waveTimer >= waveDataList[currentWaveIndex].waveDuration)
                {
                    // 开始新一波
                    StartNewWave();
                }

                if (currentWaveMonsterIndex >= nowWaveSpawnList.Count)
                {
                    waveTimer += Time.deltaTime;
                    // 当前波次怪物全部出完
                    if (isWaveInProgress)
                    {
                        EndCurrentWave();
                    }
                }
                else
                {
                    // 生成怪物
                    HandleMonsterSpawning();
                }
            }
            else
            {
                spawnedComplete = true;
            }
        }

        public virtual List<IMonster> GetAllMonsters()
        {
            return monsters;
        }

        public virtual void Init(IMapData data)
        {
            spawnedComplete = false;
            isWaveInProgress = false;
            isStarted = false;
            isPaused = true;
            currentWaveIndex = 0;
            currentWaveMonsterIndex = 0;
            waveTimer = 0;
            monsterSpawnTimer = 0;
            // 转换地图数据
            pathList = data.GetPathList();
            waveDataList = data.GetWaveData();
            obstacleList = data.GetObstacle();

            CreateObstacles();
            CreateCarrot();
            CreateStartBrand();

            // 更新面板波数显示
            sceneManger = GameManager.Instance.sceneManager as INormalSceneManager;
            sceneManger?.UpdateWaveCount(1, waveDataList.Count);

            monsters.Clear();
        }

        #region 集火相关

        public virtual IMonster GetCollectingFiresTarget()
        {
            return collectingFiresTarget;
        }

        public virtual void CancelCollectingFiresTarget()
        {
            collectingFiresTarget = null;
            // 隐藏集火标志
            signTrans.gameObject.SetActive(false);
            signTrans.transform.SetParent(transform);
        }

        #endregion

        #region 出怪相关

        public virtual void StartSpawn()
        {
            isStarted = true;
            isPaused = false;
            currentWaveIndex = 0;
            currentWaveMonsterIndex = 0;
            waveTimer = 0f;
            monsterSpawnTimer = 0f;
            isWaveInProgress = false;
            Debug.Log("Wave spawning started.");
        }

        public virtual void PauseSpawn()
        {
            isPaused = true;
        }

        public virtual void ResumeSpawn()
        {
            isPaused = false;
        }

        private void StartNewWave()
        {
            isWaveInProgress = true;
            currentWaveMonsterIndex = 0;
            waveTimer = 0f;
            monsterSpawnTimer = 0f;

            // 生成出怪数据结构list
            nowWaveSpawnList.Clear();
            foreach (EachWaveData eachWaveData in waveDataList[currentWaveIndex].eachWaveDataList)
            {
                for (int i = 0; i < eachWaveData.monsterCount; i++)
                {
                    nowWaveSpawnList.Add(new SpawnMonsterData()
                    {
                        monsterType = eachWaveData.monsterType,
                        nextSpawnTime = eachWaveData.monsterDuration,
                        hard = eachWaveData.hard
                    });
                }
            }

            // 更新面板波数显示
            GameFacade.Instance.SendNotification(NotificationName.UI.WAVES_COUNT_UPDATED, (currentWaveIndex + 1, waveDataList.Count));
        }

        private void EndCurrentWave()
        {
            isWaveInProgress = false;
            currentWaveIndex++;
        }

        private void HandleMonsterSpawning()
        {
            if (currentWaveIndex == 0 && currentWaveMonsterIndex == 0)
            {
                // 第一波的第一个直接出怪不用计时
                SpawnMonster(nowWaveSpawnList[currentWaveMonsterIndex].monsterType, nowWaveSpawnList[currentWaveMonsterIndex].hard);
                currentWaveMonsterIndex++;
                monsterSpawnTimer = 0;
            }
            else
            {
                monsterSpawnTimer += Time.deltaTime;
                if (monsterSpawnTimer >= nowWaveSpawnList[currentWaveMonsterIndex].nextSpawnTime)
                {
                    SpawnMonster(nowWaveSpawnList[currentWaveMonsterIndex].monsterType, nowWaveSpawnList[currentWaveMonsterIndex].hard);
                    currentWaveMonsterIndex++;
                    monsterSpawnTimer = 0;
                }
            }
        }

        private void SpawnMonster(EMonsterType type, float hard)
        {
            Monster monster = GameManager.Instance.poolManager.GetObject("Object/Monster/" + type).GetComponent<Monster>();
            monster.transform.SetParent(transform);
            monster.transform.localScale = Vector3.one;
            MonsterDataMap map = Resources.Load<MonsterDataMap>("Data/Monster/MonsterDataMap");
            monster.Init(pathList, hard, map.GetData(type));
            monsters.Add(monster);
        }

        #endregion

        #region 升级、出售

        /// <summary>
        /// 升级塔
        /// </summary>
        public virtual void UpGradeTower(Vector3 cellWorldPos)
        {
            ITower tower = Map.GetCell(cellWorldPos).tower as ITower;
            if (tower == null) return;
            var level = tower.GetLevel();
            var data = tower.GetData();
            // 最大等级直接返回
            if (level == 2) return;

            // 够钱才升级
            if (GameManager.Instance.sceneManager.GetMoney() < data.prices[level + 1]) return;
            // 扣钱
            sceneManger.UpdateMoney(-data.prices[level + 1]);
            // 调用更新方法
            tower.UpGrade();
            // 关闭建造面板
            GameFacade.Instance.SendNotification(NotificationName.UI.HIDE_BUILT_PANEL);
        }

        /// <summary>
        /// 出售塔
        /// </summary>
        public virtual void SellTower(Vector3 cellWorldPos)
        {
            Cell cell = Map.GetCell(cellWorldPos);
            ITower tower = cell.tower as ITower;
            if (tower == null) return;
            var level = tower.GetLevel();
            var data = tower.GetData();
            // 加钱
            sceneManger.UpdateMoney(+data.sellPrices[level]);
            // 回收对象
            GameManager.Instance.poolManager.PushObject(tower.Transform.gameObject);
            // 清空格子
            cell.tower = null;
            // 关闭建造面板
            GameFacade.Instance.SendNotification(NotificationName.UI.HIDE_BUILT_PANEL);
            // 从列表移除
            towers.Remove(tower);
        }

        #endregion

        #region 创建对象

        public virtual void SetCollectingFires(IMonster monster)
        {
            for (int i = 0; i < towers.Count; i++)
            {
                towers[i].SetCollectingFiresTarget(monster);
            }

            // 设置集火标志
            collectingFiresTarget = monster;
            signTrans.gameObject.SetActive(true);
            signTrans.SetParent(monster.GetSignFather());
            signTrans.localPosition = Vector3.zero;
            signTrans.localScale = Vector3.one;
        }

        /// <summary>
        /// 根据保存的地图数据生成障碍物
        /// </summary>
        private void CreateObstacles()
        {
            for (int i = 0; i < obstacleList.Count; i++)
            {
                if (obstacleList[i].obstacleName == "None")
                {
                    continue;
                }

                Cell cell = obstacleList[i];
                // 创建实例
                Obstacle obstacle = GameManager.Instance.poolManager.GetObject($"Object/Obstacle/{cell.obstacleName}").GetComponent<Obstacle>();
                obstacle.transform.SetParent(transform);
                obstacle.transform.localScale = Vector3.one;
                obstacle.transform.position = Map.GetCellCenterPos(cell);
                cell.obstacle = obstacle.gameObject;
                obstaclesList.Add(obstacle);
            }
        }

        /// <summary>
        /// 创建塔对象
        /// </summary>
        /// <param name="towerData"></param>
        /// <param name="cellWorldPos">创建的位置世界坐标</param>
        public virtual void CreateTowerObject(TowerData towerData, Vector3 cellWorldPos)
        {
            // 够钱才创建
            if (GameManager.Instance.sceneManager.GetMoney() >= towerData.prices[0])
            {
                ITower tower = GameManager.Instance.poolManager.GetObject(towerData.prefabsPath).GetComponent<ITower>();
                tower.Transform.SetParent(transform);
                tower.Transform.localScale = Vector3.one;
                tower.Transform.position = cellWorldPos;
                // 扣钱
                sceneManger.UpdateMoney(-towerData.prices[0]);
                // 记录该格子已经存在塔
                Map.GetCell(cellWorldPos).tower = tower;
                // 关闭建造面板
                GameFacade.Instance.SendNotification(NotificationName.UI.HIDE_BUILT_PANEL);
                // 添加进列表
                if (!towers.Contains(tower))
                {
                    towers.Add(tower);
                }
            }
        }

        /// <summary>
        /// 创建萝卜
        /// </summary>
        private void CreateCarrot()
        {
            carrot = GameManager.Instance.poolManager.GetObject("Object/Carrot").GetComponent<Carrot>();
            carrot.transform.SetParent(transform);
            carrot.transform.localScale = Vector3.one;
            // 设置萝卜位置
            Cell lastPathCell = pathList[pathList.Count - 1];
            carrot.transform.position = Map.GetCellCenterPos(lastPathCell);
        }

        /// <summary>
        /// 创建开始路牌
        /// </summary>
        private void CreateStartBrand()
        {
            startPoint = Instantiate(Resources.Load<GameObject>("Object/StartPoint")).GetComponent<Transform>();
            startPoint.SetParent(transform);
            startPoint.localScale = Vector3.one;
            // 设置开始路牌位置
            Cell firstPathCell = pathList[0];
            startPoint.position = Map.GetCellCenterPos(firstPathCell);
        }

        #endregion

        #region 缓存池相关

        /// <summary>
        /// 回收所有游戏対象
        /// </summary>
        public virtual void OnPushAllGameObject()
        {
            // 隐藏标志
            signTrans.gameObject.SetActive(false);

            GameManager.Instance.poolManager.PushObject(carrot.gameObject);
            OnPushAllTowers();
            OnPushAllMonsters();
            OnPushAllObstacles();
        }

        /// <summary>
        /// 回收未死亡的怪物
        /// </summary>
        private void OnPushAllMonsters()
        {
            for (int i = 0; i < monsters.Count; i++)
            {
                if (!monsters[i].IsDead)
                {
                    GameManager.Instance.poolManager.PushObject(monsters[i].Transform.gameObject);
                }
            }

            monsters.Clear();
        }

        private void OnPushAllTowers()
        {
            for (int i = 0; i < towers.Count; i++)
            {
                GameManager.Instance.poolManager.PushObject(((MonoBehaviour)towers[i]).gameObject);
            }

            towers.Clear();
        }

        private void OnPushAllObstacles()
        {
            for (int i = 0; i < obstaclesList.Count; i++)
            {
                Obstacle o = (Obstacle)obstaclesList[i];
                if (!o.IsDead)
                {
                    GameManager.Instance.poolManager.PushObject(o.gameObject);
                }
            }

            obstaclesList.Clear();
        }

        #endregion

        #region 外部调用

        public virtual int GetNowWaveCount()
        {
            return currentWaveIndex;
        }

        #endregion

        #region 判断胜利

        public virtual bool WinJudge()
        {
            // 1.出怪完成
            if (!spawnedComplete)
            {
                return false;
            }

            // 2.萝卜没死
            if (carrot.IsDead)
            {
                return false;
            }

            // 3.怪物全部死亡
            for (int i = 0; i < monsters.Count; i++)
            {
                // 有一个没死亡都无效
                if (monsters[i].IsDead == false)
                {
                    return false;
                }
            }

            return true;
        }

        #endregion
    }
}